### PR TITLE
add Optuna Transformers integration

### LINKED ISSUE #8: add `transformers.py`
@twolffpiggott should we add this here or to an other project we open source?

https://github.com/PhilipMay/mltb/blob/master/mltb/integration/transformers.py

### LINKED ISSUE #38: Flake8 ignore list for Black compatibility
Flake8 raises a warning for "E203" when it encounters a Black decision to insert whitespace before `:` in slicing syntax.

[Black's behaviour is more correct here](https://github.com/psf/black/issues/315), so my suggestion is to add "E203" to the flake8 config ignore list.

i.e. in `setup.cfg`:

```
[flake8]
...
extend-ignore = E203

### PR DESCRIPTION
**Pull request by twolffpiggott**
Add optuna transformers callback integration

Resolves #8 , resolves #38

### CODE REVIEW

File: LICENSE [hunk 1]
```diff
@@ -1,7 +1,8 @@
 MIT License
 
 Copyright (c) 2021 Philip May
-Copyright (c) 2021 Philip May, Deutsche Telekom AG
+Copyright (c) 2021 Timothy Wolff-Piggott
+Copyright (c) 2021 Philip May, Timothy Wolff-Piggott, Deutsche Telekom AG
```
Comment (PhilipMay) [inline LICENSE:5]: I think better keep one person per line.

File: setup.py [hunk 2]
```diff
@@ -17,7 +17,7 @@
 ]
 extras_require = {
     "checking": ["black", "flake8", "isort", "mdformat"],
-    "optional": ["mlflow", "GitPython"],
+    "optional": ["mlflow", "GitPython", "scikit-learn", "torch", "transformers"],
```
Comment (PhilipMay) [inline setup.py:20]: I think torch should be moved to "checking" because it is used in the tests.

Comment (twolffpiggott) [reply setup.py:20]: Yeah I wanted to chat about it. Should we perhaps add a "testing" option?

For dependencies specifically needed for unit tests. Whereas "checking" right now covers formatting dependencies.

File: LICENSE (hunk 1 above)
Comment (twolffpiggott) [reply LICENSE:5]: Should I update to the following:

```
Copyright (c) 2021 Philip May
Copyright (c) 2021 Philip May, Deutsche Telekom AG
Copyright (c) 2021 Timothy Wolff-Piggott
```

Are the two lines with your name distinguishing between your personal contributions and your contributions through project work for Deutsche Telekom AG?

Comment (PhilipMay) [reply LICENSE:5]: > Should I update to the following:

Yes please

> Are the two lines with your name distinguishing between your personal contributions and your contributions through project work for Deutsche Telekom AG?

Yes. Exactly.

File: setup.py (hunk 2 above)
Comment (PhilipMay) [reply setup.py:20]: > Should we perhaps add a "testing" option?

Yes please.

File: hpoflow/optuna_transformers.py [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+import os
+
+import mlflow
+
+from hpoflow.optuna_mlflow import OptunaMLflow
+from optuna._imports import try_import
+
+with try_import() as _imports:
+    import transformers
+_imports.check()
```
Comment (PhilipMay) [inline hpoflow/optuna_transformers.py:15]: I think this `_imports.check()` should not be called here but later when the class is actualy used (in the `__init__` function).

File: hpoflow/optuna_transformers.py [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+
+_logger = logging.getLogger(__name__)
+
+
+class OMLflowCallback(transformers.TrainerCallback):
+    """
+    Class based on `transformers.TrainerCallback`; integrates with OptunaMLflow
+    to send the logs to `MLflow` and `Optuna` during model training.
```
Comment (PhilipMay) [inline hpoflow/optuna_transformers.py:23]: I think this should be RST and they use 2 backticks - not just one.

So something like this:
```
    Class based on ``transformers.TrainerCallback``; integrates with OptunaMLflow
    to send the logs to ``MLflow`` and ``Optuna`` during model training.
```

Also see here for example: https://github.com/optuna/optuna/blob/99484054fcec97eb530960a8612c3e6058ee5690/optuna/pruners/_median.py#L58

File: hpoflow/optuna_transformers.py (hunk 3 above)
Comment (twolffpiggott) [reply hpoflow/optuna_transformers.py:15]: This was my original implementation but it ends up being a bit more complicated, because the `OMLflowCallback` class needs to inherit from `transformers.TrainerCallback`, which may or may not be available.

https://github.com/PhilipMay/mltb/pull/53#issuecomment-825796849

I can switch back to that original, but we previously decided to favour the simpler implementation

Comment (twolffpiggott) [reply hpoflow/optuna_transformers.py:15]: This is the lazy import implementation:

https://github.com/PhilipMay/mltb/pull/53/commits/73e9bbfd6d7de32a990607e2c7dda332097547da

Comment (PhilipMay) [reply hpoflow/optuna_transformers.py:15]: Ahh yes. I remember the discussion. Sorry. So this is ok.

**Comment by twolffpiggott** [comment]
@PhilipMay all comments resolved and ready for final review

File: CONTRIBUTING.md [hunk 5]
```diff
@@ -123,3 +123,5 @@ Contributions must be licensed under the same license as code, the [MIT license]
 
 [@PhilipMay](https://github.com/PhilipMay) - general documentation, GitHub actions,
 `optuna_mlflow.py`, `optuna.py`, everything else
+
+[@Tim Wolff-Piggott](https://github.com/twolffpiggott) - `optuna_transformers.py`, general documentation
```
Comment (PhilipMay) [inline CONTRIBUTING.md:127]: @twolffpiggott 
Can you please change 
`[@Tim Wolff-Piggott]` to `[@twolffpiggott]`

File: hpoflow/optuna_transformers.py (hunk 3 above)
Comment (PhilipMay) [reply hpoflow/optuna_transformers.py:15]: @twolffpiggott maybe add a comment that explains this? like this:

`# do the check here and not in the constructor because the class inherits from transformers.TrainerCallback`

File: hpoflow/optuna_transformers.py [hunk 6]
```diff
@@ -0,0 +1,125 @@
...
+    ):
+        """
+        Check integration package dependencies and initialize class.
+        Args:
+            trial: OptunaMLflow object
+            log_training_args: Whether to log all Transformers TrainingArguments as MLflow params
+            log_model_config: Whether to log the Transformers model config as MLflow params
+        """
+        self._MAX_PARAM_VAL_LENGTH = mlflow.utils.validation.MAX_PARAM_VAL_LENGTH
+        self._MAX_PARAMS_TAGS_PER_BATCH = mlflow.utils.validation.MAX_PARAMS_TAGS_PER_BATCH
```
Comment (PhilipMay) [inline hpoflow/optuna_transformers.py:40]: Would it be better not to create these variables but use `mlflow.utils.validation.MAX_PARAM_VAL_LENGTH` and `mlflow.utils.validation.MAX_PARAMS_TAGS_PER_BATCH` directly below?

File: hpoflow/optuna_transformers.py [hunk 7]
```diff
@@ -0,0 +1,125 @@
...
+            self.setup(args, state, model)
+        if state.is_world_process_zero:
+            metrics_to_log = dict()
+            for k, v in logs.items():
+                if isinstance(v, (int, float)):
+                    metrics_to_log[k] = v
+                else:
+                    _logger.warning(
+                        f"Trainer is attempting to log a value of "
+                        f'"{v}" of type {type(v)} for key "{k}" as a metric. '
```
Comment (PhilipMay) [inline hpoflow/optuna_transformers.py:112]: I think changing `'` and `"` signs here would be cleaner.

File: hpoflow/optuna_transformers.py (hunk 6 above)
Comment (twolffpiggott) [reply hpoflow/optuna_transformers.py:40]: I don't have a strong opinion either way. Happy to just access the variables directly.

File: tests/test_optuna_transformers.py [hunk 8]
```diff
@@ -0,0 +1,168 @@
...
+    train_func = _train_func_factory(config, output_path, model, tracking_file_name)
+    study.optimize(train_func, n_trials=N_TRIALS)
+
+    mlflow_client = MlflowClient(tracking_file_name)
+
+    # test that there is 1 MLflow experiment, corresponding to the Optuna trial
+    experiments = mlflow_client.list_experiments()
+    assert len(experiments) == 1
+    experiment = experiments[0]
+    assert experiment.name == study_name
```
Comment (PhilipMay) [inline tests/test_optuna_transformers.py:72]: Should we add `assert experiment is not None`?

File: tests/test_optuna_transformers.py [hunk 9]
```diff
@@ -0,0 +1,168 @@
...
+    tempdir = tempfile.mkdtemp()
+    return Path(tempdir)
+
+
+@pytest.fixture(scope="module")
+def tracking_file_name(output_path):
+    return f"file:{output_path.as_posix()}"
+
+
+def test_omlflow_callback(config, output_path, model, tracking_file_name):
```
Comment (PhilipMay) [inline tests/test_optuna_transformers.py:60]: Should we use the tmpdir-fixture or tmp-path-fixture from pytest here instead of `tracking_file_name`?
https://docs.pytest.org/en/6.2.x/tmpdir.html

**Comment by PhilipMay** [comment]
Many thanks @twolffpiggott 👍 
I left some inline comments. Not 100% sure about them but it would be nice if you have a look.

File: tests/test_optuna_transformers.py (hunk 8 above)
Comment (twolffpiggott) [reply tests/test_optuna_transformers.py:72]: In the scenario where the experiment is not persisted correctly, `experiments` will just be an empty list and will fail the assert on line 70.

I can't think of a case where we expect the MLflow client to return a `NoneType` experiment; do you know of one?

Comment (PhilipMay) [reply tests/test_optuna_transformers.py:72]: > do you know of one

no :-)

File: tests/test_optuna_transformers.py (hunk 9 above)
Comment (twolffpiggott) [reply tests/test_optuna_transformers.py:60]: Thanks, I've updated to use `tmp_path_factory` to create a single module-scoped tmpdir fixture. `tmp_path` is only function-scoped I think

https://pytest.org/en/latest/how-to/tmpdir.html#tmpdir

Comment (twolffpiggott) [reply tests/test_optuna_transformers.py:57]: This allows us to access the same file in both the training and test functions

**Comment by twolffpiggott** [comment]
Thanks @PhilipMay I've pushed the latest updates

File: tests/test_optuna_transformers.py (hunk 9 above)
Comment (PhilipMay) [reply tests/test_optuna_transformers.py:57]: Ok.
